### ISSUE TITLE
Photoshop guidelines and magicksave: libMagick error

### ISSUE THREAD

**Issue by kelilevi** [libvips/ruby-vips#246]
I have a few thousand .tif images created in Photoshop containing 4 guidelines inside. Using ruby-vips on Windows, I need to crop the middle portion based on the 4 guidelines. Is there an easy way to detect the PS guidelines in ruby-vips? 

Found a PS script option on: https://www.ps-scripts.com/viewtopic.php?t=9822
*the info about guides can be found in the photoshop tag hex 35 42 49 4D 04 08  the length of the tag is found at offset 11. that value is the length to the next tag the number of guides can be found at offset 27. From that point the data repeats for each guide. 4 bytes for the location. That value / 32 = pixels and one byte for type 0 = vert and 1 = horz 

Secondly: when trying to save the final image to PDF (ruby-vips with magicksave) I am getting the “magicksave: libMagick error: NoEncodeDelegateForThisImageFormat” Tried with multiple file formats, all have the same error.

Any help or pointing to the right direction is much appreciated.

**Comment by jcupitt** [comment]
Hello @kelilevi,

You'll need to parse that metadata out somehow. Do you have a sample image?

Save as PDF works on Windows at the CLI for me using:

https://github.com/libvips/libvips/releases/download/v8.10.1/vips-dev-w64-all-8.10.1.zip

```
$ ./vips.exe magicksave ~/x.jpg x.pdf
```

So it should work in ruby-vips. What version are you running?

**Comment by kelilevi** [comment]
Thank you @jcupitt for your fast respond. 

Tried libvips v8.10.1 in CLI and is working (only issue with that - I can't use with ruby and is not carrying over the profile – the output pdf is untagged and is flat looking compare to the original). If I am using the “rmagick” option all is good – same embedded profile, but is really slow and is taking forever to save with the big images:
```
#!/usr/bin/ruby
require 'vips'
include Vips
vips_cache_set_max(0)
require "rmagick"

file2open = "./2698060_kicsi.tif"
file2save = file2open.gsub('.tif', '.pdf')

img_array = Magick::Image.read(file2open)
imgs = Magick::ImageList.new
for image in img_array
    imgs<<image
end
imgs.write (file2save)

#im = Vips::Image.new_from_file file2open 
#im.magicksave(file2save)
### ...magicksave: libMagick error: NoEncodeDelegateForThisImageFormat `
```
If I try to open a .jpg with Magick::Image.read I am getting the: “Wrong JPEG library version: library is 80, caller expects 62” error… Trying to fix that too...

Do you know how I can update my ruby-vips to the “libvips v8.10.1”? Currently when I install (gem install ruby-vips) is grabbing the ruby-vips-2.0.17 with a warning: mingw-w64-x86_64-libvips-8.9.1-4 is up to date -- skipping

Attached the sample Photoshop image with guidelines, if you don’t mind, could you please point me to the right direction how I can parse out the guidelines metadata (top, bottom, left, right) in ruby-vips or ruby? And, if you know a good fast way to save the image to PDF, I'll be more than happy. Sorry for bombarding you with a lot of question, I am searching the web for days and days without any luck... 

[2698060_kicsi.zip](https://github.com/libvips/ruby-vips/files/5179102/2698060_kicsi.zip)

Thank you very much for your help!

**Comment by jcupitt** [comment]
libvips sets the imagemagick profile, so the lack of profiles must be a problem with the imagemagick PDF writer, I think.

The `Wrong JPEG library version` error comes if a program has been compiled against one version of the libjpeg headers, but is being run against the DLL for a different libjpeg. It sounds like IM has been built against the old libjpeg6.2, but is being run against the modern libjpeg 8.0. You probably need to check your `PATH`.

How did you install libvips? I would just download the new zip somewhere and set the Ruby DLL path to point to that.

I'll have a quick look at your sample image.

**Comment by jcupitt** [comment]
Argh actually ICC profiles in `magicksave` were not working. It looks like IM have changed their metadata naming conventions. I've updated it and added some extra tests and this fix should be in 8.10.2.

I had a look at your image -- I think you need to pull out the 6728 bytes of `photoshop-data`, eg.:

```
$ irb
irb(main):001:0> require 'vips'
=> true
irb(main):002:0> x = Vips::Image.new_from_file("2698060_kicsi.tif")
irb(main):004:0> data = x.get("photoshop-data")
irb(main):005:0> data.length
=> 6728
```

And then use Ruby's `unpack` method to parse the grid lines:

https://www.rubydoc.info/stdlib/core/2.0.0/String:unpack

**Comment by kelilevi** [comment]
@jcupitt, thank you for your guide, I was able to get the guidelines info:
```
#!/usr/bin/ruby
ENV["VIPS_WARNING"] = '1' # For ignoring VIPS-WARNING: Incompatible type for "RichTIFFIPTC" - for tif images
require 'vips'

im = Vips::Image.new_from_file "./image.tif"
ps_data = im.get("photoshop-data")

guideTag = '8BIM' + 4.chr + 8.chr
tagPos = ps_data.index(guideTag)
guides_count = ps_data[tagPos + 27].ord

puts "Guides_count = " + guides_count.to_s
pointer = tagPos + 28
i = 1
until i > guides_count do
   guide_type = ps_data[pointer + 4].ord > 0  ? 'horz' : 'vert'
   n = Integer('0x' + ((ps_data[pointer].ord).to_s(16) + (ps_data[pointer+1].ord).to_s(16) + (ps_data[pointer+2].ord).to_s(16) + (ps_data[pointer+3].ord).to_s(16))) / 32
   puts "Guide #{i} = " + guide_type + " = " +  (n).to_s
   pointer = pointer + 5
   i +=1
end
```
Looking forward to be able to download the vips-dev-w64-all-8.10.2 fix.
Thank you again!
